shaders/ray_queries/ray_shadow.frag

```diff
@@ -0,0 +1,121 @@
+#version 460
+#extension GL_EXT_ray_query : enable
+
+/* Copyright (c) 2021 Holochip Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 the "License";
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+layout(location = 0) in vec4 in_pos;
+layout(location = 1) in vec3 in_normal;
+layout(location = 2) in vec4 in_scene_pos;
+
+layout(location = 0) out vec4 o_color;
+
+layout(set = 0, binding = 0) uniform accelerationStructureEXT topLevelAS;
+
+layout(set = 0, binding = 1) uniform GlobalUniform
+{
+	mat4 model;
+	mat4 view_proj;
+	vec3 camera_position;
+	vec3 light_position;
+}
+global_uniform;
+
+
+/**
+Calculate ambient occlusion
+*/
+float calculate_ambient_occlusion(vec3 object_point, vec3 object_normal)
+{
+	const float ao_mult = 1;
+	uint max_ao_each = 3;
+	uint max_ao = max_ao_each * max_ao_each;
+	const float max_dist = 2;
+	const float tmin = 0.01, tmax = max_dist;
+	float accumulated_ao = 0.f;
+	vec3 u = abs(dot(object_normal, vec3(0, 0, 1))) > 0.9 ? cross(object_normal, vec3(1, 0, 0)) : cross(object_normal, vec3(0, 0, 1));
+	vec3 v = cross(object_normal, u);
+	float accumulated_factor = 0;
+	for (uint j = 0; j < max_ao_each; ++j)
+	{
+		float phi = 0.5*(-3.14159 + 2 * 3.14159 * (float(j + 1) / float(max_ao_each + 2)));
+		for (uint k = 0; k < max_ao_each; ++k){
+			float theta =  0.5*(-3.14159 + 2 * 3.14159 * (float(k + 1) / float(max_ao_each + 2)));
+			float x = cos(phi) * sin(theta);
+			float y = sin(phi) * sin(theta);
+			float z = cos(theta);
+			vec3 direction = x * u + y * v + z * object_normal;
+
+			rayQueryEXT query;
+			rayQueryInitializeEXT(query, topLevelAS, gl_RayFlagsTerminateOnFirstHitEXT, 0xFF, object_point, tmin, direction.xyz, tmax);
+			while (rayQueryProceedEXT(query))
```

Comment:
similar to `intersects_light` below, please avoid loop

Reply:
Do you have a suggestion for what to change it to?
I was under the impression that rayQueryProceedExt is meant to return false until it is finished traversing.  gl_RayFlagsTerminateOnFirstHitExp should hopefully ensure this is only called once.  If it's changed to a different flag, then the while loop might have a real cost but doesn't this combo result in a no cost loop that shows correct usage?

Reply:
ah sorry, I meant to replace it with exactly the same as in lines 100-105 below, to call `rayQueryProceed` by itself rather than in an empty while loop

Reply:
Sorry still a little confused, line 100 has an empty while loop as well.

Do you mean the nested for loop logic?  Move the while loop outside the for loop(s)?  This should run 9 times as written with a no-op essentially on the while loop as it's going to be a first hit.